client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -186,8 +191,143 @@ private Checkpoint completeCheckpoint(String checkpointName) {
+                    });
+                    val newState = synchronizer.getState();
+                    // If successfullyUpdated is true then that means the update can take place and at this point the
+                    // update is completed and it is in the desired state on which doReinit is called to continue the
+                    // reset with this newState.
+                    if (successfullyUpdated.get()) {
+                        doReinit(newState);
+                        return;
+                    }
+                    break;
+                case DELETING:
+                    doDelete(state);
+                    throw new ReinitializationRequiredException("This ReaderGroup does not exist.");
+            }
+        }
+    }
+
+    public void createState(ReaderGroupConfig config) {
+        val state = getState();
+        if (state.getConfigState() == ReaderGroupState.ConfigState.INITIALIZING && state.getConfig().equals(config)) {
+            doInit(state);
+        }
+    }
+
+    public void deleteState() {
+        while (true) {
+            val state = getState();
+            switch (state.getConfigState()) {
+                case INITIALIZING:
+                    doInit(state);
+                    break;
+                case REINITIALIZING:
+                    doReinit(state);
+                    break;
+                case READY:
+                    if (updateConfigState(ReaderGroupState.ConfigState.DELETING, state)) {
+                        val newState = getState();
+                        doDelete(newState);
+                        return;
+                    }
+                    break;
+                case DELETING:
+                    doDelete(state);
+                    return;
+            }
+        }
+    }
+
+    public void doInit(ReaderGroupState state) {
+        val config = state.getConfig();
+        val generation = state.getGeneration();
+        long segment = synchronizer.getSegmentId();
+        if (!ReaderGroupConfig.RetentionConfig.NONE.equals(config.getRetentionConfig())) {
+            Set<Stream> streams = config.getStartingStreamCuts().keySet();
+            streams.forEach(s -> getThrowingException(controller.addSubscriber(scope, s.getStreamName(), groupName + segment, generation)));
+        }
+        synchronizer.updateState((s, updates) -> {
+            if (s.equals(state)) {
+                updates.add(new ChangeConfigState(ReaderGroupState.ConfigState.READY, s.getGeneration()));
+            }
+        });
+    }
+
+    public void doReinit(ReaderGroupState state) {
+        val oldConfig = state.getConfig();
+        val newConfig = state.getNewConfig();
+        val generation = state.getGeneration();
+        manageSubscriptions(oldConfig, newConfig, generation);
+
+        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
+        synchronizer.updateState((s, updates) -> {
+            if (s.equals(state)) {
```

Comment:
same here: repeating what i said earlier:

we should do a specific field comparison rather than compare the whole state object.
the reason is - state object can be modified even by other updates like checkpoints etc. 
so what we really care about is if the state.generation and state.oldconfig state.newconfig and state.status are identical

Reply:
Addressed above.